Sid/edit docker docs (#565)
voxel51/fiftyone-teams-app-deploy: docker/README.md

- [:white_check_mark: Technical Requirements](#white_check_mark-technical-requirements)
- [:clock10: Estimated Completion Time](#clock10-estimated-completion-time)
- [:floppy_disk: Sizing](#floppy_disk-sizing)

> - [:wrench: Step 0: Set Up MongoDB Database](#wrench-step-0-set-up-mongodb-database)

Reviewer: Isn't "Step 0" really "Step 1"? Or a prerequisite?

Numbering the steps like this is hard.  This is why for ordered lists, we use the markdown convention of everything being `1.` so we can reorder, add new and not have to change numbers.

Does having the steps numbered help? If not, let's consider removing them.

- [:closed_lock_with_key: Step 1: Prepare License File](#closed_lock_with_key-step-1-prepare-license-file)

> - [:file_folder: Step 2: Choose Authentication Mode](#file_folder-step-2-choose-authentication-mode)

Reviewer: Why are we removing the Choose Authentication Mode step? Internal mode only supports OIDC (not SAML).  If the customer has a SAML IdP, they cannot use internal mode.

Author: ^thats for non-airgapped (standard) customers right? 

so internal should be 
1. all airgapped customers 
2. standard deployments using OIDC as IDP

legacy should be 
1. standard deployments using SAML as IDP?

Reviewer: Yes. And Air gapped only supports OIDC.  So customers should have to choose their auth mode.

  - [1. Copy the template `.env` file](#1-copy-the-template-env-file)
  - [2. Fill out required values in `.env`](#2-fill-out-required-values-in-env)
  - [3. Create a `compose.override.yaml` to override configuration](#3-create-a-composeoverrideyaml-to-override-configuration)
  - [:package: Official Docker Images](#package-official-docker-images)

> - [:rocket: Step 4: Initial Deployment](#rocket-step-3-initial-deployment)

Reviewer: the text contains `Step 4` but the link contains `step 3`?  The numbering should be the same an match the heading.  This should be managed by the pre-commit hook and not manually.

We need to run the hook again:

```shell
[fiftyone-teams-app-deploy]$ pre-commit run markdown-toc --files docker/README.md
[INFO] Initializing environment for https://github.com/syntaqx/git-hooks.
Insert a table of contents in Markdown files, like a README.md.......................Failed
- hook id: markdown-toc
- files were modified by this hook
```

  - [1. Enable Database Admin mode](#1-enable-database-admin-mode)
  - [2. Launch the application](#2-launch-the-application)
- [:globe_with_meridians: Step 5: Configure SSL & Reverse Proxy (Nginx / Load Balancer)](#globe_with_meridians-step-4-configure-ssl--reverse-proxy-nginx--load-balancer)
  - [:compass: Routing Overview (Path-Based Proxy)](#compass-routing-overview-path-based-proxy)
  - [:open_file_folder: Nginx Configuration](#open_file_folder-nginx-configuration)
  - [:memo: Notes](#memo-notes)
- [:page_facing_up: Step 6: Configure Delegated Operation Logs](#page_facing_up-step-5-configure-delegated-operation-logs)

> - [Step 6: Identity Provider (IdP) and Authentication (CAS) (Air Gapped Deployments Only)](#step-6-identity-provider-idp-and-authentication-cas-air-gapped-deployments-only)

Reviewer: Should we have two two step 6s?

Aha! We need to run the hook again:

```shell
[fiftyone-teams-app-deploy]$ pre-commit run markdown-toc --files docker/README.md
[INFO] Initializing environment for https://github.com/syntaqx/git-hooks.
Insert a table of contents in Markdown files, like a README.md.......................Failed
- hook id: markdown-toc
- files were modified by this hook
```

Resource usage varies dramatically with operations, use cases,
and dataset sizes.

## :wrench: Step 0: Set Up MongoDB Database

Before deploying FiftyOne Enterprise, you must have a running MongoDB database.
FiftyOne Enterprise supports:

- **MongoDB Atlas** (managed cloud)
- **MongoDB Community Edition** (self-hosted, open source)
- **MongoDB Enterprise** (self-hosted, commercial)

Ensure your MongoDB version meets FiftyOne's

> [version constraints](https://docs.voxel51.com/user_guide/config.html#using-a-different-mongodb-version).

Reviewer: If we link to this doc, let's work to update it to have the recommendation which is MongoDB 8.0+.

Once your database is running, record your MongoDB connection URI. You will
need it in Step 2 when configuring your `.env` file. The URI follows this

> format:

Reviewer: Aloha prefers one sentence per line.

Why?

* Consistency with the majority of the rest of the docs
* The native markdown code is easier to read
* Updating a word stands our in a diff
* Sentences spanning more than two lines are candidates for refactoring into simpler sentences

```suggestion
Once your database is running, record your MongoDB connection URI.
You will need it in Step 2 when configuring your `.env` file.
The URI follows this format:
```

> picked up immediately, you may need to restart the `teams-cas` and `teams-api`
> services.

## :gear: Step 2: Configure Environment

> Navigate into the `internal-auth` directory:

Reviewer: Repeating the comment from above:

> Why are we removing the Choose Authentication Mode step? Internal mode only supports OIDC (not SAML).  If the customer has a SAML IdP, they cannot use internal mode.

> picked up immediately, you may need to restart the `teams-cas` and `teams-api`
> services.

## :file_folder: Step 3: Choose Authentication Mode

FiftyOne Enterprise supports two authentication modes.
Choose the one that matches your deployment:

| Mode            | Use when                                                                                |
|-----------------|-----------------------------------------------------------------------------------------|
| `internal-auth` | All air-gapped deployments, OR standard deployments using an **OIDC** Identity Provider |

> | `legacy-auth`   | Standard deployments using a **SAML** Identity Provider                                 |

Reviewer: ```suggestion
| Mode            | Use when                                                                   |
|-----------------|----------------------------------------------------------------------------|
| `internal-auth` | Deployment is Air-gapped or Identity Provider is **OpenID Connect (OIDC)** |
| `legacy-auth`   | Identity Provider is **SAML**                                              |```

### 2. Launch the application

FiftyOne Enterprise is deployed with **Dedicated Plugins** and **Delegated
Operators** enabled by default. Dedicated Plugins run in an isolated
`teams-plugins` container, and Delegated Operators offload plugin execution to

> **worker containers**, enabling scalable and reliable long-running operations.

Reviewer: Sentence per line.  Split into multiple statements:

```suggestion
FiftyOne Enterprise is deployed with **Dedicated Plugins** and
**Delegated Operators** enabled by default.
Dedicated Plugins run in an isolated `teams-plugins` container.
Delegated Operators offload plugin execution to **worker containers**,
enabling scalable and reliable long-running operations.
```

> Delegated Operators offload plugin execution?
Is that true? Or is it that long running operations (that may be plugins). 

But a reader, I'd be curious why I would need Delegated Operators _and_ plugin containers. Wouldn't DO containers supersede plugin containers?

```shell
docker compose \

>   -f compose.dedicated-plugins.yaml \

Reviewer: Yes! Let's be opinionated about the default installation.

# Expected: HTTP/1.1 200 OK
```

> For a full health assessment, see [Basic Health Assessment](#basic-health-assessment).

Reviewer: ```suggestion
For a full health assessment, see
[Basic Health Assessment](#basic-health-assessment).
```

| `/_pymongo`          | `teams-api` | MongoDB requests via SDK             |
| `/health`            | `teams-api` | Health check endpoint                |

### :open_file_folder: Nginx Configuration

All services are routed based on URL path using a single hostname. See the full

> example configuration:

Reviewer: ```suggestion
All services are routed based on URL path using a
single hostname via path based routing.
See a full example configuration in
```

- To validate your deployments api connection, see
  [Validating Your Deployment](../docs/validating-deployment.md)

## :page_facing_up: Step 5: Configure Delegated Operation Logs

Add the log path for delegated operation runs to your `compose.override.yaml`:

```yaml
services:
  teams-do:
    environment:
      FIFTYONE_DELEGATED_OPERATION_LOG_PATH: "gs://your-bucket/logs"
```

Logs are stored in the format:

```text
/mnt/shared/logs/do_logs/<YYYY>/<MM>/<DD>/<RUN_ID>.log
```

This is useful for auditing, debugging, or monitoring delegated operator
executions in shared storage or cloud buckets.

<!-- markdownlint-disable-next-line MD013 -->

> ## Step 6: Identity Provider (IdP) and Authentication (CAS) (Air Gapped Deployments Only)

Reviewer: Why are we adding air gapped only here?

Internal auth mode requires setting an admin like https://docs.voxel51.com/enterprise/pluggable_auth.html#getting-started-with-internal-mode instructs. That applies to Airgapped and non-airgapped deployments. Maybe we should remove that link from this section?

   to bring up services.
1. Ensure your proxy (e.g., nginx) forwards `/cas` to the CAS service port.

## Step 7: Initial CAS Setup

Log into the CAS Super Admin UI at
`https://<ENVIRONMENT>.fiftyone.ai/cas/configurations`. When prompted, enter
the value of `FIFTYONE_AUTH_SECRET` from your `.env` file in the **API Key**

> field.

Reviewer: The user is never prompted.

```suggestion
1. Navigate to the CAS Super Admin UI at
  `https://<ENVIRONMENT>.fiftyone.ai/cas/configurations`.
1. In the **API Key** field (upper right corner), enter the
   value of `FIFTYONE_AUTH_SECRET` (from your `.env` file)
```

### Add First Admin User

1. Navigate to the **Admins** tab at `https://<ENVIRONMENT>.fiftyone.ai/cas/admins`.
1. Select **Add admin**.
1. Provide **Name** and **Email**.
1. Select **Add**.

### Enable Auto Join

1. Navigate to `https://<ENVIRONMENT>.fiftyone.ai/cas/providers`.
1. Select **+ Edit**.
1. Select **Allow auto join**.
1. Select **Save**.

## Step 8: Test End User Login

> Verify the deployment is working end-to-end by logging in as a regular user.

Reviewer: Is logging in, really an end-to-end verification?

Author: hmm @kevin-dimichel  i think its just a sanity check that they can move on to the next step

Reviewer: Yes. But let's not call it an end-to-end when it isn't : D

Please update.

> 1. Open `https://<ENVIRONMENT>.fiftyone.ai` in a browser.

Reviewer: ```suggestion
1. In a browser, open `https://<ENVIRONMENT>.fiftyone.ai`.
```

> 1. Log in with the admin user credentials you created in Step 7.

Reviewer: But step 7 didn't create credentials, just the admin user

```suggestion
1. Log in with the user credentials of the admin you created in Step 7.
```

1. Confirm you are redirected to the FiftyOne Enterprise home page.

## Advanced DO Settings

### :desktop_computer: GPU-Enabled Workloads

FiftyOne services like Delegated Operators can be scheduled on **GPU-enabled

> hardware** for more efficient computation.

Reviewer: Don't line split in the middle of bolded text.
```suggestion
FiftyOne services like Delegated Operators can be scheduled on
**GPU-enabled hardware** for more efficient computation.
```

To set up containers with GPU resources, see the
[configuring GPU workloads documentation](./docs/configuring-gpu-workloads.md).

### :bricks: Custom Plugin Images

If your delegated operators or plugins require **custom dependencies**, build
and deploy **custom plugin images**. Base them on
`voxel51/fiftyone-teams-cv-full`, which includes a full CV/ML environment, and

> extend with:

Reviewer: ```suggestion
If your delegated operators or plugins require **custom dependencies**, build
and deploy **custom plugin images**.
Use `voxel51/fiftyone-teams-cv-full` as the base image
(that includes a full CV/ML environment), and extend with:
```

- Custom Python packages
- Internal SDKs or models

### :on: On-Demand Delegated Operator Executors

FiftyOne Enterprise v2.11 introduces support for on-demand delegated operator
executors for Databricks and Anyscale. Please refer to the
[configuration documentation](https://github.com/voxel51/fiftyone-teams-app-deploy/blob/main/helm/docs/configuring-delegated-operators.md).

> 

Reviewer: ```suggestion
FiftyOne Enterprise v2.11 introduces support for on-demand
delegated operator executors for Databricks and Anyscale.
Please refer to the
[configuration documentation](https://github.com/voxel51/fiftyone-teams-app-deploy/blob/main/helm/docs/configuring-delegated-operators.md).

```

## Upgrades

The recommended upgrade path is:

1. Pull the latest compose files from this repo:

   ```shell
   git pull origin main
   ```

>    This automatically populates the latest service versions.

Reviewer: Let's be consistent in our block quoting (to match the next step)

```suggestion
   > This automatically populates the latest service versions.
```

1. Confirm `FIFTYONE_DATABASE_ADMIN` is set to `false` in your
   `compose.override.yaml`:

   ```yaml
   services:
     fiftyone-app:
       environment:
         FIFTYONE_DATABASE_ADMIN: false
   ```

   > This prevents automatic database migrations from running on startup and
   > breaking active SDK sessions.

1. If you are using [Custom Plugin Images](#bricks-custom-plugin-images),
   rebuild them using the updated base image version, then update their tags

>    in `compose.override.yaml` to match the new release.

Reviewer: ```suggestion
1. When using [Custom Plugin Images](#bricks-custom-plugin-images),
   rebuild them using the updated base image version.
   Update their tags in `compose.override.yaml` to match the new release.
```

1. Bring the stack down and back up:

   ```shell
   docker compose \
     -f compose.dedicated-plugins.yaml \
     -f compose.delegated-operators.yaml \
     -f compose.override.yaml \
     down

   docker compose \
     -f compose.dedicated-plugins.yaml \
     -f compose.delegated-operators.yaml \
     -f compose.override.yaml \
     up -d

>    ```

Reviewer: ````suggestion
1. Bring the compose stack down

   ```shell
   docker compose \
     -f compose.dedicated-plugins.yaml \
     -f compose.delegated-operators.yaml \
     -f compose.override.yaml \
     down
   ```

1. Bring the new stack up:

   ```shell
   docker compose \
     -f compose.dedicated-plugins.yaml \
     -f compose.delegated-operators.yaml \
     -f compose.override.yaml \
     up -d
   ```
````

   to bring up services.
1. Ensure your proxy (e.g., nginx) forwards `/cas` to the CAS service port.

## Step 7: Initial CAS Setup

1. Navigate to the CAS Super Admin UI at
  `https://<ENVIRONMENT>.fiftyone.ai/cas/configurations`.
1. In the **API Key** field (upper right corner), enter the
   value of `FIFTYONE_AUTH_SECRET` (from your `.env` file)

### Add First Admin User

1. Navigate to the **Admins** tab at `https://<ENVIRONMENT>.fiftyone.ai/cas/admins`.
1. Select **Add admin**.
1. Provide **Name** and **Email**.
1. Select **Add**.

### Enable Auto Join

1. Navigate to `https://<ENVIRONMENT>.fiftyone.ai/cas/providers`.
1. Select **+ Edit**.
1. Select **Allow auto join**.
1. Select **Save**.

## Step 8: Test End User Login

Verify the deployment is working end-to-end by logging in as a regular user.

1. In a browser, open `https://<ENVIRONMENT>.fiftyone.ai`.

> 1. Log in with the user credentials of the admin you created in Step 7.

Reviewer: Let's make this less brittle:

```suggestion
1. Log in with the user credentials of the admin
   you created in the CAS Super Admin UI
```

   to bring up services.
1. Ensure your proxy (e.g., nginx) forwards `/cas` to the CAS service port.

## Step 9: Initial CAS Setup

1. Navigate to the CAS Super Admin UI at
  `https://<ENVIRONMENT>.fiftyone.ai/cas/configurations`.
1. In the **API Key** field (upper right corner), enter the
   value of `FIFTYONE_AUTH_SECRET` (from your `.env` file)

### Add First Admin User

1. Navigate to the **Admins** tab at `https://<ENVIRONMENT>.fiftyone.ai/cas/admins`.
1. Select **Add admin**.
1. Provide **Name** and **Email**.
1. Select **Add**.

### Enable Auto Join

1. Navigate to `https://<ENVIRONMENT>.fiftyone.ai/cas/providers`.
1. Select **+ Edit**.
1. Select **Allow auto join**.
1. Select **Save**.

## Step 10: Test End User Login

> Verify the deployment's IDP setup by logging in as a regular user.

Reviewer: ```suggestion
Verify the deployment's IdP setup by logging in as a regular user.
```

1. In a browser, open `https://<ENVIRONMENT>.fiftyone.ai`.
1. Log in with the user credentials of the admin
   you created in the CAS Super Admin UI.
1. Confirm you are redirected to the FiftyOne Enterprise home page.

## Advanced DO Settings

### :desktop_computer: GPU-Enabled Workloads

FiftyOne services like Delegated Operators can be scheduled on
**GPU-enabled hardware** for more efficient computation.

To set up containers with GPU resources, see the
[configuring GPU workloads documentation](./docs/configuring-gpu-workloads.md).

### :bricks: Custom Plugin Images

If your delegated operators or plugins require **custom dependencies**, build
and deploy **custom plugin images**.
Use `voxel51/fiftyone-teams-cv-full` as the base image
(that includes a full CV/ML environment), and extend with:

- Custom Python packages
- Internal SDKs or models

Once built, reference your custom image in `compose.override.yaml`:

```yaml
services:
  teams-do:
    image: fiftyone-cv-full-custom:v2.19.0
```

### :on: On-Demand Delegated Operator Executors

FiftyOne Enterprise v2.11 introduces support for on-demand
delegated operator executors for Databricks and Anyscale.
Please refer to the

> [configuration documentation](https://github.com/voxel51/fiftyone-teams-app-deploy/blob/main/helm/docs/configuring-delegated-operators.md).

Reviewer: ```suggestion
[configuring delegated operators documentation](https://github.com/voxel51/fiftyone-teams-app-deploy/blob/main/helm/docs/configuring-delegated-operators.md).
```

## Upgrades

The recommended upgrade path is:

1. Pull the latest compose files from this repo:

   ```shell
   git pull origin main

>    ```

Reviewer: ````suggestion
    ```shell
    git pull origin main
    ```
````

   > This automatically populates the latest service versions.

1. Confirm `FIFTYONE_DATABASE_ADMIN` is set to `false` in your
   `compose.override.yaml`:

   ```yaml
   services:
     fiftyone-app:
       environment:
         FIFTYONE_DATABASE_ADMIN: false

>    ```

Reviewer: ````suggestion
    ```yaml
    services:
      fiftyone-app:
        environment:
          FIFTYONE_DATABASE_ADMIN: false
    ```
````

   > This prevents automatic database migrations from running on startup and
   > breaking active SDK sessions.

1. When using [Custom Plugin Images](#bricks-custom-plugin-images),
   rebuild them using the updated base image version.
   Update their tags in `compose.override.yaml` to match the new release.

1. Bring the compose stack down:

   ```shell
   docker compose \
     -f compose.dedicated-plugins.yaml \
     -f compose.delegated-operators.yaml \
     -f compose.override.yaml \
     down

>    ```

Reviewer: ````suggestion
    ```shell
    docker compose \
      -f compose.dedicated-plugins.yaml \
      -f compose.delegated-operators.yaml \
      -f compose.override.yaml \
      down
    ```
````

1. Bring the new stack up:

   ```shell
   docker compose \
     -f compose.dedicated-plugins.yaml \
     -f compose.delegated-operators.yaml \
     -f compose.override.yaml \
     up -d

>    ```

Reviewer: ````suggestion
    ```shell
    docker compose \
      -f compose.dedicated-plugins.yaml \
      -f compose.delegated-operators.yaml \
      -f compose.override.yaml \
      up -d
    ```
````
